### PR TITLE
refactor(conductor): adapt executor to sequential firm, soft blocks

### PR DESCRIPTION
**Pull request by SuperFluffy**
## Summary
Refactor executor to the stricter contract, expecting firm (from celestia) and soft (from sequencer) blocks to be in strict sequential order.

## Background
Enforcing a contract that the celestia and sequencer readers only forward blocks in strict sequential order of heights makes the executor implementation much simpler.

## Changes
- executor expects to receive firm/soft blocks in strict sequential order
- soft blocks are dropped if they are too old
- old firm blocks, or soft/firm blocks that lie in the future violate the contract between the executor and {celestia, sequencer}-readers and lead to conductor aborting
- move storage of blocks pending finalization into the `execute_soft` handler

## Testing
The executor tests have been updated to support various scenarios of receiving old blocks, out of order blocks, new blocks etc.

## Related Issues
Part of https://github.com/astriaorg/astria/pull/691

### CODE REVIEW

File: crates/astria-conductor/src/executor/track_state.rs [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+/// to `uint64` heights by the times this becomes an issue.
+fn map_rollup_height_to_sequencer_height(
+    first_sequencer_height: u32,
+    current_rollup_height: u32,
+) -> Height {
+    first_sequencer_height
+        .checked_add(current_rollup_height)
+        .expect(
+            "should not overflow; if this overflows either the first sequencer height or current \
+             rollup height have been incorrectly set incorrectly set, are in the future, or the \
```
Comment (sambukowski) [inline crates/astria-conductor/src/executor/track_state.rs:20]: nit: duplicate "incorrectly set"

**Review by sambukowski — APPROVED** [review]

File: crates/astria-conductor/src/block_cache.rs [hunk 2]
```diff
@@ -29,9 +29,9 @@ pub(crate) struct BlockCache<T> {
 impl<T> BlockCache<T> {
     /// Creates a new block cache.
     ///
-    /// The first block it will serve is at height 0.
+    /// The default first block is at height 1.
     pub(crate) fn new() -> Self {
-        Self::with_next_height(Height::from(0u32))
+        Self::with_next_height(Height::default())
```
Comment (noot) [inline crates/astria-conductor/src/block_cache.rs:34]: maybe just have an explicit height 1?

File: crates/astria-conductor/src/executor/track_state.rs [hunk 3]
```diff
@@ -0,0 +1,159 @@
+use astria_core::execution::v1alpha2::{
+    Block,
+    CommitmentState,
+};
+use sequencer_client::tendermint::block::Height;
+
+// Maps a rollup height to a sequencer heights.
+/// # Panics
```
Comment (noot) [inline crates/astria-conductor/src/executor/track_state.rs:8]:
```suggestion
/// Maps a rollup height to a sequencer heights.
///
/// # Panics
```

File: crates/astria-core/src/sequencer/v1alpha1/test_utils.rs [hunk 4]
```diff
@@ -57,25 +59,39 @@ impl ConfigureCometBftBlock {
...
+        // let rollup_id = RollupId::from_unhashed_bytes([b"test_chain_id_", &*suffix].concat());
+        let unsigned_transaction = UnsignedTransaction {
+            nonce: 1,
+            // actions: vec![
+            //     SequenceAction {
+            //         rollup_id,
+            //         data: [b"hello_world_id_", &*suffix].concat(),
+            //     }
+            //     .into(),
+            // ],
```
Comment (noot) [inline crates/astria-core/src/sequencer/v1alpha1/test_utils.rs:93]: remove commented out stuff

**Review by noot — APPROVED** [review]

File: crates/astria-conductor/src/executor/tests.rs [hunk 5]
```diff
@@ -201,110 +240,191 @@ async fn start_mock_with_optimism_handler() -> MockEnvironmentWithEthereum {
...
-    // deletes it from the map
-    assert!(mock.executor.sequencer_hash_to_execution_block.is_empty());
 }
 
 #[tokio::test]
-async fn try_execute_out_of_order_block_from_sequencer() {
+async fn old_soft_blocks_are_ignored() {
+    let mut mock = start_mock(None).await;
+    let mut block = make_cometbft_block();
+    block.header.height = Height::from(99u32);
```
[the remark this answers is not in the document]
Comment (SuperFluffy) [reply crates/astria-conductor/src/executor/tests.rs:386]: added it to the assert msg

File: crates/astria-conductor/src/block_cache.rs (hunk 2 above)
Comment (SuperFluffy) [reply crates/astria-conductor/src/block_cache.rs:34]: Removed the `BlockCache::new` constructor and return an error if `BlockCache::with_next_height` is called with a height of zero.

File: crates/astria-conductor/src/executor/track_state.rs (hunk 3 above)
Comment (SuperFluffy) [reply crates/astria-conductor/src/executor/track_state.rs:8]: removed in the meantime.

**Pull request merged**